EDU-18422 Data layer & data fetching - Sales App Extensibility (#2698)
vtexdocs/dev-portal-content: docs/guides/VTEX-Sales-App/data-layer-and-data-fetching-in-vtex-sales-app-extensibility.md

---
title: "Data layer and data fetching in VTEX Sales App Extensibility"
slug: "data-layer-and-data-fetching-in-vtex-sales-app-extensibility"
hidden: false
excerpt: "Learn how to use the data layer and data fetching for Sales App extensions."
createdAt: "2026-05-27T00:00:00.000Z"
---

> ⚠️ This feature is in beta, and we're working to improve it. If you have any questions, please contact our [Support](https://help.vtex.com/en/support).

> When using [VTEX Sales App Extensibility](link-placeholder), you may need to present data to the user, whether it’s data from the Sales App flow itself, VTEX APIs, or third-party APIs.

Reviewer: ```suggestion
When [creating a Sales App extension](https://developers.vtex.com/docs/guides/creating-sales-app-extensions), you may need to present data to the user, whether it's from the Sales App flow, VTEX APIs, or third-party APIs.
```

> In extension points, there are two ways to interact with data:

Reviewer: ```suggestion
There are two ways for enabling your extensions to interact with data:
```

- Data layer resources, such as the `useCart` hook or `useCartItem`.
- Data fetching from VTEX APIs or external APIs.

## Data layer in extensions

All interactions with the Sales App data layer happen through functions and hooks provided by the `@vtex/sales-app` package.

> > ℹ️ Whenever possible, we recommend using the Sales App data layer, because this data is already cached in the core data layer. By doing so, you won’t trigger additional requests, which improves application performance and benefits your extensions.

Reviewer: ```suggestion
> ℹ️ Whenever possible, use the Sales App data layer, because this data is already cached in the core data layer. This approach prevents additional requests, enhancing application performance and benefiting your extensions.
```

For example, if you need to access cart item data while using the `cart.cart-item.after` extension point, you can use the `useCartItem` hook:

> ```typescript

Reviewer: ```suggestion
```tsx
```

import { useCartItem } from '@vtex/sales-app'

const MyComponent = () => {
  const { item } = useCartItem()

  return <p>Quantity: {item.quantity}</p>
}
```

> > ℹ️ For detailed information about hooks and utilities for extensions, see the [API Reference](link-placeholder).

Reviewer: ```suggestion
> ℹ️ For detailed information about hooks and types for extensions, see the guide [Sales App extensions hooks and types](https://developers.vtex.com/docs/guides/sales-app-extensions-hooks-and-types).
```

> ## Data fetching in extensions

Reviewer 2: Talvez vale adicionar um disclaimer orientando que não é seguro fazer chamadas diretas caso precise passar token de autenticação para a API (não é o caso do exemplo), pois o token ficaria exposto.
Para esse caso que citei, é aconselhável criar um app io que servirá de proxy para a chamada e faz essa autenticação segura.

> When you need to fetch data from VTEX APIs (e.g., Intelligent Search) or external APIs, you can use the browser's Fetch API to make requests, as shown in the example below:

Reviewer: ```suggestion
When you need to fetch data from VTEX APIs (for example, Intelligent Search) or external APIs, you can use the browser's Fetch API to make requests, as shown in the example below:
```

```tsx
import React, { useState, useEffect } from 'react';

function MyCustomData() {
  const [response, setResponse] = useState({ status: 'loading' });

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch('https://my-custom-data-api.com/custom-data');
        const data = await response.json();
        setResponse({ status: 'data', data });
      } catch (error) {
        setResponse({ status: 'error' })
      }
    };
    fetchData();
  }, []);

  if (response.status === 'loading') return <MySkeleton />
  if (response.status === 'error') return null

  return <p>Data: {data.information}</p>
}
```

> > ⚠️ When performing data fetching, it’s important to handle loading states to ensure a good user experience. Additionally, allocate space in advance to avoid layout shift. Read more about [Dealing with Layout Shift](link-placeholder).

Reviewer: ```suggestion
> ⚠️ When fetching data, handle loading states to ensure a better user experience. Additionally, allocate space in advance to avoid layout shift. Learn more in the guide [Exploring Sales App extensions](https://developers.vtex.com/docs/guides/exploring-sales-app-extensions).
```
